File: apps/ios/Domain/Tests/DomainTests/EntitlementsVectors.swift
import Foundation
@testable import Domain

// P1.7b: wires the real Entitlements.swift port into FunctionRegistry so
// entitlements.json's vectors un-skip. Mirrors Android's
// EntitlementsVectors.kt.

func registerEntitlementsVectors() {
    let domain = "entitlements"

    FunctionRegistry.register(domain: domain, fn: "canUse") { input in
        let d = input as! [String: Any]
        return canUse(d["feature"] as! String, d["tier"] as! String)
    }
    FunctionRegistry.register(domain: domain, fn: "isPremiumFeature") { input in
        let d = input as! [String: Any]
        return isPremiumFeature(d["feature"] as! String)
    }

    FunctionRegistry.register(domain: domain, fn: "entitlementState") { input in
        let d = input as! [String: Any]
        func str(_ k: String) -> String? { d[k] as? String }
        func int(_ k: String) -> Int? { (d[k] as? NSNumber)?.intValue }
        let s = entitlementState(
            tier: str("tier"),
            premiumTrialStartDate: str("premiumTrialStartDate"),
            compTier: str("compTier"),
            compUntil: str("compUntil"),
            nowMillis: (d["nowMillis"] as! NSNumber).int64Value,
            monthlyQuotaTotal: int("monthlyQuotaTotal"),
            monthlyQuotaUsed: int("monthlyQuotaUsed"),
            purchasedQuotaRemaining: int("purchasedQuotaRemaining"),
            additionalPurchasedQuota: int("additionalPurchasedQuota")
        )
        return [
            "tier": s.tier,
            "isPaid": s.isPaid,
            "isTrial": s.isTrial,
            "trialDaysLeft": s.trialDaysLeft,
            "quotaTotal": s.quotaTotal,
            "quotaUsed": s.quotaUsed,
            "purchased": s.purchased,
            "quotaLeft": s.quotaLeft,
        ] as [String: Any]
    }
}

File: apps/android/domain/src/test/kotlin/com/sanvya/app/domain/entitlements/EntitlementsVectors.kt
package com.sanvya.app.domain.entitlements

import com.sanvya.app.domain.vectors.FunctionRegistry
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.long
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

// P1.7a: wires the real Entitlements.kt port into FunctionRegistry so
// entitlements.json's vectors un-skip.

fun registerEntitlementsVectors() {
    val domain = "entitlements"

    FunctionRegistry.register(domain, "canUse") { input ->
        val o = input.jsonObject
        JsonPrimitive(canUse(o.getValue("feature").jsonPrimitive.content, o.getValue("tier").jsonPrimitive.content))
    }
    FunctionRegistry.register(domain, "isPremiumFeature") { input ->
        JsonPrimitive(isPremiumFeature(input.jsonObject.getValue("feature").jsonPrimitive.content))
    }

    FunctionRegistry.register(domain, "entitlementState") { input ->
        val o = input.jsonObject
        fun str(k: String): String? = o[k]?.jsonPrimitive?.contentOrNull
        fun int(k: String): Int? = o[k]?.jsonPrimitive?.contentOrNull?.toIntOrNull()
        val s = entitlementState(
            tier = str("tier"),
            premiumTrialStartDate = str("premiumTrialStartDate"),
            compTier = str("compTier"),
            compUntil = str("compUntil"),
            nowMillis = o.getValue("nowMillis").jsonPrimitive.long,
            monthlyQuotaTotal = int("monthlyQuotaTotal"),
            monthlyQuotaUsed = int("monthlyQuotaUsed"),
            purchasedQuotaRemaining = int("purchasedQuotaRemaining"),
            additionalPurchasedQuota = int("additionalPurchasedQuota"),
        )
        JsonObject(
            mapOf(
                "tier" to JsonPrimitive(s.tier),
                "isPaid" to JsonPrimitive(s.isPaid),
                "isTrial" to JsonPrimitive(s.isTrial),
                "trialDaysLeft" to JsonPrimitive(s.trialDaysLeft),
                "quotaTotal" to JsonPrimitive(s.quotaTotal),
                "quotaUsed" to JsonPrimitive(s.quotaUsed),
                "purchased" to JsonPrimitive(s.purchased),
                "quotaLeft" to JsonPrimitive(s.quotaLeft),
            )
        )
    }
}
